8338023: Support two vector selectFrom API

jatin-bhateja opened the pull request:
Hi All,

As per the discussion on panama-dev mailing list[1], patch adds the support for following new two vector permutation APIs.

```
Declaration:-
    Vector<E>.selectFrom(Vector<E> v1, Vector<E> v2)
```

Semantics:-
    Using index values stored in the lanes of "this" vector, assemble the values stored in first (v1) and second (v2) vector arguments. Thus, first and second vector serves as a table, whose elements are selected based on index value vector. API is applicable to all integral and floating-point types.  The result of this operation is semantically equivalent to expression v1.rearrange(this.toShuffle(), v2). Values held in index vector lanes must lie within valid two vector index range [0, 2*VLEN) else an IndexOutOfBoundException is thrown.  

Summary of changes:
-  Java side implementation of new selectFrom API.
-  C2 compiler IR and inline expander changes.
-  In absence of direct two vector permutation instruction in target ISA, a lowering transformation dismantles new IR into constituent IR supported by target platforms. 
-  Optimized x86 backend implementation for AVX512 and legacy target.
-  Function tests covering new API.

JMH micro included with this patch shows around 10-15x gain over existing rearrange API :-
Test System: Intel(R) Xeon(R) Platinum 8480+ [ Sapphire Rapids Server]

```
  Benchmark                                     (size)   Mode  Cnt      Score   Error   Units
SelectFromBenchmark.rearrangeFromByteVector     1024  thrpt    2   2041.762          ops/ms
SelectFromBenchmark.rearrangeFromByteVector     2048  thrpt    2   1028.550          ops/ms
SelectFromBenchmark.rearrangeFromIntVector      1024  thrpt    2    962.605          ops/ms
SelectFromBenchmark.rearrangeFromIntVector      2048  thrpt    2    479.004          ops/ms
SelectFromBenchmark.rearrangeFromLongVector     1024  thrpt    2    359.758          ops/ms
SelectFromBenchmark.rearrangeFromLongVector     2048  thrpt    2    178.192          ops/ms
SelectFromBenchmark.rearrangeFromShortVector    1024  thrpt    2   1463.459          ops/ms
SelectFromBenchmark.rearrangeFromShortVector    2048  thrpt    2    727.556          ops/ms
SelectFromBenchmark.selectFromByteVector        1024  thrpt    2  33254.830          ops/ms
SelectFromBenchmark.selectFromByteVector        2048  thrpt    2  17313.174          ops/ms
SelectFromBenchmark.selectFromIntVector         1024  thrpt    2  10756.804          ops/ms
SelectFromBenchmark.selectFromIntVector         2048  thrpt    2   5398.244          ops/ms
SelectFromBenchmark.selectFromLongVector        1024  thrpt    2   5856.859          ops/ms
SelectFromBenchmark.selectFromLongVector        2048  thrpt    2   1513.378          ops/ms
SelectFromBenchmark.selectFromShortVector       1024  thrpt    2  17888.617          ops/ms
SelectFromBenchmark.selectFromShortVector       2048  thrpt    2   9079.565          ops/ms

```
Kindly review and share your feedback.

Best Regards,
Jatin

[1] https://mail.openjdk.org/pipermail/panama-dev/2024-May/020408.html

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8340338](https://bugs.openjdk.org/browse/JDK-8340338) to be approved
- [x] Change must be properly reviewed (3 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 2 [Authors](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8338023](https://bugs.openjdk.org/browse/JDK-8338023): Support two vector selectFrom API (**Enhancement** - P4)
 * [JDK-8340338](https://bugs.openjdk.org/browse/JDK-8340338): Support two vector selectFrom API (**CSR**)


### Reviewers
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20508/head:pull/20508` \
`$ git checkout pull/20508`

Update a local copy of the PR: \
`$ git checkout pull/20508` \
`$ git pull https://git.openjdk.org/jdk.git pull/20508/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20508`

View PR using the GUI difftool: \
`$ git pr show -t 20508`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20508.diff">https://git.openjdk.org/jdk/pull/20508.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20508#issuecomment-2276277689)

jatin-bhateja commented:
/label add hotspot-compiler-dev

PaulSandoz reviewed (commented):
The results look promising. I can provide guidance on the specification e.g., we can specify the behavior in terms of rearrange, with the addition of throwing on out of bounds indexes.

Regarding the throwing of exceptions, some wider context will help to know where we are heading before we finalize the specification. I believe we are considering changing the default throwing behavior for index out of bounds to wrapping, thereby we can avoid bounds checks. If that is the case we should wait until that is done then update rather than submitting a CSR just yet?

I see you created a specific intrinsic, which will avoid the cost of shuffle creation. Should we apply the same approach (in a subsequent PR) to the single argument shuffle? Or perhaps if we manage to optimize shuffles and change the default wrapping we don't require a specific intrinsic and can just use defer to rearrange?

jatin-bhateja commented:
> The results look promising. I can provide guidance on the specification e.g., we can specify the behavior in terms of rearrange, with the addition of throwing on out of bounds indexes.
> 
> Regarding the throwing of exceptions, some wider context will help to know where we are heading before we finalize the specification. I believe we are considering changing the default throwing behavior for index out of bounds to wrapping, thereby we can avoid bounds checks. If that is the case we should wait until that is done then update rather than submitting a CSR just yet?
> 
> I see you created a specific intrinsic, which will avoid the cost of shuffle creation. Should we apply the same approach (in a subsequent PR) to the single argument shuffle? Or perhaps if we manage to optimize shuffles and change the default wrapping we don't require a specific intrinsic and can just use defer to rearrange?

Hi @PaulSandoz ,
Thanks for your comments. With this new API we intend to enforce stricter specification w.r.t to index values to emit a lean instruction sequence preventing any cycles spent on massaging inputs to a consumable form and preventing redundant wrapping and unwrapping operations.

Existing [two vector rearrange API](https://docs.oracle.com/en/java/javase/22/docs/api/jdk.incubator.vector/jdk/incubator/vector/Vector.html#rearrange(jdk.incubator.vector.VectorShuffle,jdk.incubator.vector.Vector)) has a flexible specification which allows wrapping out of bounds shuffle indexes into exceptional index with a -ve value.

Even if we optimize existing two vector rearrange implementation we will still need to emit additional instructions to generate an indexes which lie within two vector range [0, 2*VLEN). I see this as a specialized API like vector compress/expand which cater to targets like x86-AVX512+ and aarch64-SVE which offers direct instruction for two vector lookups.

May be the API nomenclature can be refined to better reflect its semantics i.e. from selectFrom to twoVectorLookup ?

turbanoff commented on test/jdk/jdk/incubator/vector/IntMaxVectorTests.java:
```diff
@@ -318,6 +318,25 @@ static void assertexpandArraysEquals(int[] r, int[] a, boolean[] m, int vector_l
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/FloatMaxVectorTests.java:
```diff
@@ -335,6 +335,25 @@ static void assertexpandArraysEquals(float[] r, float[] a, boolean[] m, int vect
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Float256VectorTests.java:
```diff
@@ -330,6 +330,25 @@ static void assertexpandArraysEquals(float[] r, float[] a, boolean[] m, int vect
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Byte128VectorTests.java:
```diff
@@ -313,6 +313,25 @@ static void assertexpandArraysEquals(byte[] r, byte[] a, boolean[] m, int vector
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Double64VectorTests.java:
```diff
@@ -330,6 +330,25 @@ static void assertexpandArraysEquals(double[] r, double[] a, boolean[] m, int ve
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Float512VectorTests.java:
```diff
@@ -330,6 +330,25 @@ static void assertexpandArraysEquals(float[] r, float[] a, boolean[] m, int vect
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Long64VectorTests.java:
```diff
@@ -270,6 +270,25 @@ static void assertexpandArraysEquals(long[] r, long[] a, boolean[] m, int vector
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/DoubleMaxVectorTests.java:
```diff
@@ -335,6 +335,25 @@ static void assertexpandArraysEquals(double[] r, double[] a, boolean[] m, int ve
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Short256VectorTests.java:
```diff
@@ -313,6 +313,25 @@ static void assertexpandArraysEquals(short[] r, short[] a, boolean[] m, int vect
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Long256VectorTests.java:
```diff
@@ -270,6 +270,25 @@ static void assertexpandArraysEquals(long[] r, long[] a, boolean[] m, int vector
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Float128VectorTests.java:
```diff
@@ -330,6 +330,25 @@ static void assertexpandArraysEquals(float[] r, float[] a, boolean[] m, int vect
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

turbanoff commented on test/jdk/jdk/incubator/vector/Int512VectorTests.java:
```diff
@@ -313,6 +313,25 @@ static void assertexpandArraysEquals(int[] r, int[] a, boolean[] m, int vector_l
+            int idx = i + j;
+            boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+            int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " +  b[i + oidx]);
```
```suggestion
            Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]), "at index #" + idx + ", order = " + (int)order[idx] + ", a = " + a[i + oidx] + ", b = " + b[i + oidx]);
```

jatin-bhateja commented:
> _Mailing list message from [John Rose](mailto:john.r.rose@oracle.com) on [hotspot-compiler-dev](mailto:hotspot-compiler-dev@mail.openjdk.org):_
> 
> (Better late than never, although I wish I?d been more explicit about this on panama-dev.)
> 
> I think we should be moving away from throwing exceptions on all reorder/shuffle/permute vector ops, and moving toward wrapping. These ops all operate on vectors (small arrays) of vector lane indexes (small array indexes in a fixed domain, always a power of two). The throwing behavior checks an input for bad indexes and throws a (scalar) exception if there are any at all. The wrapping behavior reduces bad indexes to good ones by an unsigned modulo operation (which is at worst a mask for powers of two).
> 
> If I?m right, then new API points should start out with wrap semantics, not throw semantics. And old API points should be migrated ASAP.
> 
> There?s no loss of functionality in such a move. Instead the defaults are moved around. Before, throwing was the default and wrapping was an explicit operation. After, wrapping would be the default and throwing would be explicit. Both wrapping and throwing checks are available through explicit calls to VectorShuffle methods checkIndexes and wrapIndexes.
> 
> OK, so why is wrapping better than throwing? And first, why did we start with throwing as the default? Well, we chose throwing as the default to make the vector operations more Java-like. Java scalar operations don?t try to reduce bad array indexes into the array domain; they throw. Since a shuffle op is like an array reference, it makes sense to emulate the checks built into Java array references.
> 
> Or it did make sense. I think there is a technical debt here which is turning out to be hard to pay off. The tech debt is to suppress or hoist or strength-reduce the vector instructions that perform the check for invalid indexes (in parallel), then ask ?did any of those checks fail?? (a mask reduction), then do a conditional branch to failure code. I think I was over-confident that our scalar tactics for reducing array range checks would apply to vectors as well. On second thought, vectorizing our key optimization, of loop range splitting (pre/main/post loops) is kind of a nightmare.
> 
> Instead, consider the alternative of wrapping. First, you use vpand or the like to mask the indexes down to the valid range. Then you run the shuffle/permute instruction. That?s it. There is no scalar query or branch. And, there are probably some circumstances where you can omit the vpand operation: Perhaps the hardware already masks the inputs (as with shift instructions). Or, perhaps C2 can do bitwise inference of the vectors and figure out that the vpand is a nop. (I am agitating for bitwise types in C2; this is a use case for them.) In the worst case, the vpand op is fast and pipelines well.
> 
> This is why I think we should switch, ASAP, to masking instead of throwing, on bad indexes.
> 
> I think some of our reports from customers have shown that the extra checks necessary for throwing on bad indexes are giving their code surprising slowdowns, relative to C-based vector code.
> 
> Did I miss a point?
> 
> ? John
> 
> On 14 Aug 2024, at 3:43, Jatin Bhateja wrote:


Hi @rose00,

I agree that wrapping should be the default behaviour if indices are passed through shuffles, idea was to pick exception throwing semantics for out of bounds indexes *only* for selectFrom flavour of APIs which accept indexes through vector interface, this will save redundant partial wrapping and un-wrapping for cross vector permutation API which has a direct mappings in x86 and AARCH64 ISA.

As @PaulSandoz [suggested](https://github.com/openjdk/jdk/pull/20508#pullrequestreview-2234095541) we can also tune existing single 'selectFrom' API to adopt default exception throwing semantics if any of the indices lies beyond valid index range.

While we will continue keeping default wrapping semantics for APIs accepting shuffles, this little deviation of semantics for selectFrom family of APIs will enable generating efficient code and will enable users to chooses between the rearrange and selectFrom APIs based on convenience vs efficient code trade-off.

Since, API interfaces were crafted keeping in view long term flexibility, having multiple permutation interfaces (selectFrom / rearrange) accepting indexes though vector or shuffle enables compiler to emit efficient code.

Best Regards,
Jatin

sviswa7 commented:
@rose00 @PaulSandoz Please see the work in progress (https://github.com/openjdk/jdk/pull/20634) to make wrap indices as default for rearrange and selectFrom apis.

jatin-bhateja commented:
Hi @rose00 , @sviswa7 , @PaulSandoz , 
As suggested, now passing explicit 'wrap' argument to new selectFrom API.

Following are the performance number of modified JMH micro included with the patch.

```

Baseline:-
Benchmark                                      (size)   Mode  Cnt     Score   Error   Units
SelectFromBenchmark.rearrangeFromByteVector     4096  thrpt    2  5849.771          ops/ms
SelectFromBenchmark.rearrangeFromDoubleVector   4096  thrpt    2   430.712          ops/ms
SelectFromBenchmark.rearrangeFromFloatVector    4096  thrpt    2   942.737          ops/ms
SelectFromBenchmark.rearrangeFromIntVector      4096  thrpt    2  1057.695          ops/ms
SelectFromBenchmark.rearrangeFromLongVector     4096  thrpt    2   616.360          ops/ms
SelectFromBenchmark.rearrangeFromShortVector    4096  thrpt    2  2146.465          ops/ms

With Patch:-
Benchmark                                   (size)   Mode  Cnt     Score   Error   Units
SelectFromBenchmark.selectFromByteVector        4096  thrpt    2  9543.775          ops/ms
SelectFromBenchmark.selectFromDoubleVector      4096  thrpt    2   558.195          ops/ms
SelectFromBenchmark.selectFromFloatVector       4096  thrpt    2  1325.059          ops/ms
SelectFromBenchmark.selectFromIntVector         4096  thrpt    2  1418.748          ops/ms
SelectFromBenchmark.selectFromLongVector        4096  thrpt    2   687.231          ops/ms
SelectFromBenchmark.selectFromShortVector       4096  thrpt    2  4782.395          ops/ms


With WIP wrap index acceleration PR#20634:
Benchmark                                      (size)   Mode  Cnt     Score   Error   Units
SelectFromBenchmark.rearrangeFromByteVector     4096  thrpt    2  7602.645          ops/ms
SelectFromBenchmark.rearrangeFromDoubleVector   4096  thrpt    2   441.684          ops/ms
SelectFromBenchmark.rearrangeFromFloatVector    4096  thrpt    2   926.112          ops/ms
SelectFromBenchmark.rearrangeFromIntVector      4096  thrpt    2  1061.695          ops/ms
SelectFromBenchmark.rearrangeFromLongVector     4096  thrpt    2   644.058          ops/ms
SelectFromBenchmark.rearrangeFromShortVector    4096  thrpt    2  2777.735          ops/ms

```

sviswa7 commented:
@jatin-bhateja Thanks, the PR ((https://github.com/openjdk/jdk/pull/20634) is still work in progress and can be simplified much further. The changes I am currently working on are do wrap by default for rearrange and selectFrom as suggested by John and Paul, no additional api with boolean wrap as parameter, and no changes to shuffle constructors.

PaulSandoz commented:
Is it possible for the intrinsic to be responsible for wrapping, if needed? If was looking at [`vpermi2b`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=vpermi2b&ig_expand=4917,4982,5004,5010,5014&techs=AVX_512) and AFAICT it implicitly wraps, operating on the lower N bits. Is that correct?

sviswa7 commented:
> Is it possible for the intrinsic to be responsible for wrapping, if needed? If was looking at [`vpermi2b`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=vpermi2b&ig_expand=4917,4982,5004,5010,5014&techs=AVX_512) and AFAICT it implicitly wraps, operating on the lower N bits. Is that correct?

It is good to keep wrapping separate. Two reasons: 1) Not all permute instructions do wrapping e.g. pshufb has a different behavior if MSB is set. 2) By keeping wrapping separate it can move out of the loop if shuffle is loop invariant.

sviswa7 commented:
@rose00 @PaulSandoz I have updated https://github.com/openjdk/jdk/pull/20634. Please take a look if it meets your expectations for the existing rearrange/selectFrom apis. Jatin can then base the new two vector selectFrom api in this PR on similar lines.

jatin-bhateja commented:
Hi @rose00 , @PaulSandoz , @sviswa7,

Latest patch removed explicit wrap argument passed to selectFrom API, instead uses wrapping scheme as a mitigation strategy to handle OOB partially wrapped indexes. 

Summarizing the new scheme for index wrapping:-
 - Shuffle always holds indexes in valid vector index range.
 - Following are the shuffle creation intercepts 
       - VectorShuffle.fromArray -  Partially wraps OOB indexes
       - iotaShuffle                      -  Accepts explicit wrap argument to chooses b/w wrapping vs partial wrapping of OOB indexes.
       - Vector.toShuffle              - Partially wraps OOB indexes.   
- Partial wrapping generate -ve indexes for OOB indices after wrapping them into valid index range.
      - Objective is to delegate mitigation strategy to subsequent APIs which can either generate a IndexOutOfBounds exception or  create valid index by adding vector length.
     -  An important point to mention here is that partially wrapped indexing schemes first wraps OOB index  ( index < 0 OR  index >= VECLEN) into valid index range and then subtracts VECLEN from wrapped index to generate a -ve number in [-VECLEN: -1] range.
- With new scheme we are choosing wrapping as a default mitigation strategy hence only client which make effective use of a partially wrapped indexes is two vector re-arrange, which uses it to compute the mask for blending two permuted vectors. 
-  Two vector re-arrange and selectFrom API differ in terms of acceptable index range, while former accepts shuffle indices in single vector index range [0:VECLEN) latter operates on two vector index range [0:2*VECLEN).

Best Regards,
Jatin

PaulSandoz commented:
API changes look good. (Note at the moment we are not proposing to change how shuffles works - as you point out the two vector `selectFrom` and `rearrange` differ in the index representation.)

IIUC if the more direct two-table instruction is not available you fall back to calling two single arg rearranges with a blend, as a lowering transformation, similar to the fallback Java expression.

The float/double conversion bothers me, not suggesting we do something about it here, noting down for any future conversion on shuffles. Ideally we would want the equivalent integral vector (int or long) to represent the index, tricky to express in the API, or alternative treat as a bitwise no-op conversion (there is also impact on `toShuffle` too).

jatin-bhateja commented:
> API changes look good. (Note at the moment we are not proposing to change how shuffles works - as you point out the two vector `selectFrom` and `rearrange` differ in the index representation.)
> 
> IIUC if the more direct two-table instruction is not available you fall back to calling two single arg rearranges with a blend, as a lowering transformation, similar to the fallback Java expression.
> 
> The float/double conversion bothers me, not suggesting we do something about it here, noting down for any future conversation on shuffles. Ideally we would want the equivalent integral vector (int or long) to represent the index, tricky to express in the API, or alternative treat as a bitwise no-op conversion (there is also impact on `toShuffle` too).

Thanks @PaulSandoz, 

> IIUC if the more direct two-table instruction is not available you fall back to calling two single arg rearranges with a blend, as > a lowering transformation, similar to the fallback Java expression.

Idea here is to be performant as much as possible and save additional boxing penalties incurred due to failed intrinsification if target does not directly support two vector permutation but does supports its constituents. I have now unwrapped and optimized the fallback implementation to directly operates over index vector lanes instead going through intermediate shuffle.

> 
> The float/double conversion bothers me, not suggesting we do something about it here, noting down for any future conversation on shuffles. 

I Agree.

> Ideally we would want the equivalent integral vector (int or long) to represent the index, tricky to express in the API, or alternative treat as a bitwise no-op conversion (there is also impact on `toShuffle` too).
 
Since floating-point index vector may carry special values like NaN, POSITIVE_INFINITY and NEGATIVE_INFINITY,  thus with default wrapping semantics, its necessary to convert this into integral vector followed by wrapping normalization to valid two vector index range,  through existing sequence we are bypassing partial wrapping altogether which may save few instruction.

PaulSandoz commented:
I think we should leave the fallback expression as `vec2.rearrange(vec1.toShuffle(), vec3);`, lets address that separately if needed. Otherwise, you have introduced an additional code path that requires more explicit testing.

My comment was related to understanding what `SelectFromTwoVectorNode::Ideal` and `VectorRearrangeNode::Ideal` are doing - the former lowers, if needed, into the rearrange expression and the latter adjusts, if needed, the index vector (a comment describing this transformation would be useful, like you have in the former method).

jatin-bhateja commented:
> I think we should leave the fallback expression as `vec2.rearrange(vec1.toShuffle(), vec3);`, lets address that separately if needed. Otherwise, you have introduced an additional code path that requires more explicit testing.
> 
> My comment was related to understanding what `SelectFromTwoVectorNode::Ideal` and `VectorRearrangeNode::Ideal` are doing - the former lowers, if needed, into the rearrange expression and the latter adjusts, if needed, the index vector (a comment describing this transformation would be useful, like you have in the former method).

Done.

eme64 commented on src/hotspot/cpu/x86/x86.ad:
```diff
@@ -10468,3 +10485,17 @@ instruct DoubleClassCheck_reg_reg_vfpclass(rRegI dst, regD src, kReg ktmp, rFlag
 %}
+
+
+instruct selectFromTwoVec_evex(vec dst, vec src1, vec src2)
```
You could rename `dst` -> `mask_and_dst`. That would maybe help the reader to more quickly know that it is an input-mask and output-dst.

eme64 commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -2631,6 +2631,90 @@ bool LibraryCallKit::inline_vector_extract() {
+  set_result(vbox);
+  C->set_max_vector_size(MAX2(C->max_vector_size(), (uint)(num_elem * type2aelembytes(elem_bt))));
+  return true;
+}
```
The code in these methods are extremely duplicated. Unboxing and boxing in every method around here. Maybe not your problem in this PR.

BTW: your error logging used `v1` in all 3 cases `op1-3`, you probably want to give them useful names. `v1-3` probably?

All this copy-pasting makes it easy to miss updating some cases... like it happenend here.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // Lower the IR to constituents operations.
+  //   SelectFromTwoVectorNode =
+  //     (VectorBlend
+  //         (VectorRearrange SRC1, INDEX)
+  //         (VectorRearrange SRC2, NORM_INDEX)
```
```suggestion
  //         (VectorRearrange SRC1 INDEX)
  //         (VectorRearrange SRC2 NORM_INDEX)
```
Either consistently use commas or none at all ;)

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+Node* SelectFromTwoVectorNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  int num_elem = vect_type()->length();
+  BasicType elem_bt = vect_type()->element_basic_type();
+  if (Matcher::match_rule_supported_vector(Op_SelectFromTwoVector, num_elem, elem_bt)) {
```
```suggestion
  // Keep the node if it is supported, else lower it to other nodes.
  if (Matcher::match_rule_supported_vector(Op_SelectFromTwoVector, num_elem, elem_bt)) {
```

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+    return nullptr;
+  }
+  Node* index_vec = in(1);
+  Node* src1  = in(2);
+  Node* src2  = in(3);
```
```suggestion
  Node* src1 = in(2);
  Node* src2 = in(3);
```
unnecessary spaces

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // This shall prevent an intrinsification failure and associated argument
+  // boxing penalties.
+
+  auto lane_count_type = [&]() {
+    switch(elem_bt) {
+      case T_BYTE:
+      case T_SHORT:
+      case T_INT:
+      case T_FLOAT:
+        return static_cast<const Type*>(TypeInt::make(num_elem));
+      case T_DOUBLE:
+      case T_LONG:
+        return static_cast<const Type*>(TypeLong::make(num_elem));
+      default:
+        fatal("Unsupported vectortype (%s)", type2name(elem_bt));
+        return static_cast<const Type*>(nullptr);
+    }
+  };
+
+  auto make_integral_index_vec = [&](Node* index_vec) {
+    switch(elem_bt) {
+      case T_FLOAT:
+        return phase->transform(new VectorCastF2XNode(index_vec, TypeVect::make(T_INT, num_elem)));
+      break;
+      case T_DOUBLE:
+        return phase->transform(new VectorCastD2XNode(index_vec, TypeVect::make(T_LONG, num_elem)));
+      break;
+      default:
+        return index_vec;
+    }
+  };
+
+  auto get_integal_type = [&](BasicType elem_bt) {
+    switch(elem_bt) {
+      case T_FLOAT:  return T_INT;
+      case T_DOUBLE: return T_LONG;
+      default: return elem_bt;
+    }
+  };
```
This is definitely a style question. But it might be nice to make these functions member functions. They now kinda disrupt the flow of the `::Ideal` method. And in some cases you use the captured variables, and in other cases you pass them in explicitly, even though they already exist in the captured scope... consistency would be nice.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  int sopc = VectorNode::opcode(opc, integral_elem_bt);
+
+  BoolTest::mask pred = BoolTest::lt;
+  ConINode* pred_node = (ConINode*)phase->makecon(TypeInt::make(pred));
```
Would `as_ConI()` be a better alternative to the `(ConINode*)` cast?

eme64 replied:
Please at least add a comment why you are not following my suggestion. I feel like the work I put in to review is not being respected when comments are just silently resolved without any action or comment.

eme64 replied:
I really do think that `as_ConI()` would be the right thing here. In product it is just a cast, and in debug at least we have an assert.

jatin-bhateja replied:
DONE

jatin-bhateja replied:
It just got overlooked @eme64, we respect reviewer suggestions and value the time you invest in polishing our patches, thanks again :-)

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+
+  BoolTest::mask pred = BoolTest::lt;
+  ConINode* pred_node = (ConINode*)phase->makecon(TypeInt::make(pred));
+  Node* lane_cnt = phase->makecon(lane_count_type());
```
Hmm. I don't like to have different names for the same thing. `num_elem` and `lane_count` and `lane_cnt`.

What about a method `make_num_elem_node`, returns a `ConNode*`.
Then you pass it around as `num_elem_scalar`, and broadcast it to `num_elem_vector`.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+    switch(elem_bt) {
+      case T_FLOAT:
+        return phase->transform(new VectorCastF2XNode(index_vec, TypeVect::make(T_INT, num_elem)));
+      break;
```
`break` after `return`?

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+
+  // Comparison over integral vectors weeds out emitting additional
+  // instructions for checking special floating point values.
+  const TypeVect* vmask_type = TypeVect::makemask(integral_elem_bt, num_elem);
+  Node* mask = phase->transform(new VectorMaskCmpNode(pred, integral_index_vec, bcast_lane_cnt_vec, pred_node, vmask_type));
+
+  vmask_type = TypeVect::makemask(elem_bt, num_elem);
+  mask = phase->transform(new VectorMaskCastNode(mask, vmask_type));
```
I would just have two variables, and not overwrite it: `integral_vmask_type` and `vmask_type`. Maybe also `mask` could be split into two variables?

jatin-bhateja replied:
I think the variable names are appropriate and in accordance with convention.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  //         (VectorRearrange SRC2, NORM_INDEX)
+  //         MASK)
+  // This shall prevent an intrinsification failure and associated argument
+  // boxing penalties.
```
A quick comment about how the mask is computed could be nice.
`MASK = INDEX < num_elem`

eme64 replied:
@jatin-bhateja very nice, thanks!

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+      }
+    };
+    // Targets emulating unsupported permutation for certain vector types
+    // may need to message the indexes to match the users intent.
```
```suggestion
    // may need to massage the indexes to match the users intent.
```

eme64 replied:
This optimization for now seems quite specific to your `SelectFromTwoVectorNode::Ideal` lowering code. Can this conversion not be done there already?

What is the semantics of `VectorRearrangeNode`? Should its shuffle vector always be bytes, and we now violated that "for a quick second"? Or is it going to be generally the idea to create all sorts of shuffle types and then fix that up? But then why do we need the `vector_indexes_needs_massaging`?

Can you help me understand the concept/strategy behind this?

jatin-bhateja replied:
Ok, IIRC variable index permutation instruction on every target expects shape conformance b/w data vector and permute index vector.  Rearrange expects indices to be passed throug shuffle,  idealization routines automatically injects a VectorLoadShuffle after loading indexes held in shuffle's backing storage i.e. a byte array. 

In all the cases apart from byte vector permute , VectorLoadShuffle expands the index byte lanes to match the data vector lane. So we always end up emitting a lane expansion instruction before permute instruction (scenario 1). Apart from usual expansions VectorLoadShuffle may also do additional magic for some targets where it may need to prune / massage the index vector if target does not support destination vector type (scenario 2).

For our case, new selectFrom accepts the indices though vectors which save redundant expansions, but to leverage existing backend support for scenario 2 we do target specific pruning

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+      in(2)->Opcode() != Op_VectorLoadShuffle &&
+      Matcher::match_rule_supported_vector(Op_VectorRearrange, num_elem, elem_bt) &&
+      Matcher::vector_indexes_needs_massaging(elem_bt, num_elem)) {
+    auto get_integal_type = [&](BasicType elem_bt) {
+      switch(elem_bt) {
+        case T_FLOAT:  return T_INT;
+        case T_DOUBLE: return T_LONG;
+        default: return elem_bt;
+      }
+    };
```
You are now using this twice. Is there not some method that already does this?

eme64 commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -1607,11 +1608,23 @@ class VectorRearrangeNode : public VectorNode {
+class SelectFromTwoVectorNode : public VectorNode {
+public:
+  SelectFromTwoVectorNode(Node* in1, Node* in2, Node* in3, const TypeVect* vt)
+  : VectorNode(in1, in2, in3, vt) {}
```
I would prefer more expressive variable names and a short specification what the node does. Otherwise one always has to reverse-engineer what inputs are acceptable etc. I mean you could even require `VectorNode*` as inputs.

eme64 reviewed (commented):
I left a few comments, hopefully I can spend some more time on this next week

eme64 commented on test/jdk/jdk/incubator/vector/ShortMaxVectorTests.java:
```diff
@@ -1010,6 +1041,12 @@ public Object[][] shortTernaryOpProvider() {
+    public Object[][] shortSelectFromTwoVectorOpProvider() {
+        return SHORT_GENERATOR_SELECT_FROM_TRIPLES.stream().map(List::toArray).
+                toArray(Object[][]::new);
+    }
```
Just a control question: does this also occasionally generate examples with out-of-bounds indices? Negative out of bounds and positive out of bounds?

jatin-bhateja replied:
Original API did throw IndexOutOfBoundsException, but later on we have moved away from exception throwing semantics to wrapping semantics.
Please find details at following comment
https://github.com/openjdk/jdk/pull/20508#issuecomment-2306344606

jatin-bhateja replied:
VectorAPI Jtreg framework is based on testNG, our custom data providers associated with various test methods ensure to generates range of values which are beyond valid index range, this should check the wrapping scenarios.

eme64 commented on test/jdk/jdk/incubator/vector/ShortMaxVectorTests.java:
```diff
@@ -5760,6 +5797,24 @@ static void SelectFromShortMaxVectorTests(IntFunction<short[]> fa,
+                ShortVector av = ShortVector.fromArray(SPECIES, a, i);
+                ShortVector bv = ShortVector.fromArray(SPECIES, b, i);
+                ShortVector idxv = ShortVector.fromArray(SPECIES, idx, i);
+                idxv.selectFrom(av, bv).intoArray(r, i);
```
Would this test catch a bug where the backend would generate vectors that are too long or too short?

jatin-bhateja replied:
Existing vectorAPI inline expansion entry points explicitly pass lane type and count as intrinsic arguments, this is used to create concrete ideal vector types.

jatin-bhateja replied:
Patch includes tests for all the species (combination of vector type and sizes), each vector kernel is validated against equivalent scalar implementation, scenario which you are referring gets implicitly handled though tests.

eme64 replied:
Ok, just so that I can relax, can you please point me to this test that would implicitly verify that the backend has chosen the correct vector size?

jatin-bhateja replied:
Each test method validates the intrinsic code against equivalent scalar implementation, it should catch if backend emits instruction with incorrect vector size.
https://github.com/openjdk/jdk/pull/20508/files#diff-95c582657bf90bef3530e67cb143865d070fd2e8e4538849e3dce6061b0d5f2dR4863

jatin-bhateja commented on test/jdk/jdk/incubator/vector/Byte128VectorTests.java:
```diff
@@ -313,6 +313,25 @@ static void assertexpandArraysEquals(byte[] r, byte[] a, boolean[] m, int vector
+            for (; i < a.length; i += vector_len) {
+                for (j = 0; j < vector_len; j++) {
+                    int idx = i + j;
+                    boolean is_exceptional_idx = (int)order[idx] >= vector_len;
+                    int oidx = is_exceptional_idx ? ((int)order[idx] - vector_len) : (int)order[idx];
+                    Assert.assertEquals(r[idx], (is_exceptional_idx ? b[i + oidx] : a[i + oidx]));
```
I agree, but somehow we are using non camelCase conventions in this file, look for uses of 'vector_len'.  just preserving file level convention.

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/Vector.java:
```diff
@@ -2766,6 +2766,29 @@ public abstract VectorMask<E> compare(VectorOperators.Comparison op,
     public abstract Vector<E> selectFrom(Vector<E> v);
 
+    /**
+     * Rearranges the lane elements of two vectors, selecting lanes
```
We already have another flavor of [selectFrom](https://docs.oracle.com/en/java/javase/22/docs/api/jdk.incubator.vector/jdk/incubator/vector/Vector.html#selectFrom(jdk.incubator.vector.Vector)) which permutes single vector, new API extents its semantics to two vector selection, so we kept the nomenclature consistent.

PaulSandoz replied:
Select operates only on vectors where the `this` vector represents the indexes to *select* elements from the other vectors. Rearrange operates on vectors and a shuffle argument that *rearranges* elements from the other vectors. The former behavior can be specified in terms of the latter behavior, and ideally the equivalent expressions should result in ~same generated sequence of instructions. However, we are not there yet and need to further optimize shuffles to make that happen. But, we can optimize `selectFrom` with the dependent change to wrap indexes instead of throwing when out of bounds.

(Separately there is an annoying issue with select, that we should not address in this PR. Using a Float/Double Vector for indexes is awkward.)

eme64 commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:
```diff
@@ -536,6 +536,19 @@ static ByteVector compressHelper(Vector<Byte> v, VectorMask<Byte> m) {
+        byte[] res = new byte[vlen];
+        byte[] vpayload1 = ((ByteVector)v1).vec();
+        byte[] vpayload2 = ((ByteVector)v2).vec();
+        byte[] vpayload3 = ((ByteVector)v3).vec();
```
You only gave me a thumbs up and no change - but comment resolved. Is that intentional? Makes me feel like you are ignoring my comments, and that discourages me from reviewing in the future.

jatin-bhateja replied:
Routine was renamed as per you suggestion and first vector argument also appropriately renamed to wrappedIndex.

eme64 commented:
Can you please **define** somewhere what it means to `prune indexes`? It does not help me much more than the previous "massaging indexes" you had before I asked you to change it.

> Also: I'm a little worried about the semantics change of the RearrangeNode that you did with the changes in RearrangeNode::Ideal. It looks a little "hacky", especially in conjunction with the vector_indexes_needs_massaging method. Can you give a clear definition of the semantics of RearrangeNode and vector_indexes_needs_massaging, please?

You have also not responded to this yet. It seems to me that before your proposed change, `RearrangeNode` had a clear and easy semantic, and now you somehow "hack it" to work with your `vector_indexes_needs_pruning`. Can you explain please why this makes sense and add a comment to `RearrangeNode` what its semantics is?

jatin-bhateja commented:
> Can you please **define** somewhere what it means to `prune indexes`? It does not help me much more than the previous "massaging indexes" you had before I asked you to change it.
> 
> > Also: I'm a little worried about the semantics change of the RearrangeNode that you did with the changes in RearrangeNode::Ideal. It looks a little "hacky", especially in conjunction with the vector_indexes_needs_massaging method. Can you give a clear definition of the semantics of RearrangeNode and vector_indexes_needs_massaging, please?
> 
> You have also not responded to this yet. It seems to me that before your proposed change, `RearrangeNode` had a clear and easy semantic, and now you somehow "hack it" to work with your `vector_indexes_needs_pruning`. Can you explain please why this makes sense and add a comment to `RearrangeNode` what its semantics is?

In case target does not directly support two vector selection instruction we lower the IR to its constituents, this is better than  intrinsification failure as it saves costly vector boxing penalties. 

Think in terms of desired compiler IR and not rearrange API semantics,  VectorRearrange IR node always expects a shape conformance b/w vector to be permuted and index vector, since shuffle indices are held in byte array based backing storage hence compiler injects VectorLoadShuffle nodes to upcast the byte vector lanes holding indexes to match the input vector lane. 
Since selectFrom API already passes the indexes through vector hence we can save emitting redundant toShuffle() + toVector() operations in all cases apart from some target specific scenarios e.g. AVX2 targets [do not support direct short vector permute]instruction "VPERMW", hence we need to [massage the index vector](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L8771) to emulate desired permutation using byte permute instruction.

eme64 commented:
> > Can you please **define** somewhere what it means to `prune indexes`? It does not help me much more than the previous "massaging indexes" you had before I asked you to change it.
> > > Also: I'm a little worried about the semantics change of the RearrangeNode that you did with the changes in RearrangeNode::Ideal. It looks a little "hacky", especially in conjunction with the vector_indexes_needs_massaging method. Can you give a clear definition of the semantics of RearrangeNode and vector_indexes_needs_massaging, please?
> > 
> > 
> > You have also not responded to this yet. It seems to me that before your proposed change, `RearrangeNode` had a clear and easy semantic, and now you somehow "hack it" to work with your `vector_indexes_needs_pruning`. Can you explain please why this makes sense and add a comment to `RearrangeNode` what its semantics is?
> 
> In case target does not directly support two vector selection instruction we lower the IR to its constituents, this is better than intrinsification failure as it saves costly vector boxing penalties.
> 
> Consider in terms of desired compiler IR and not rearrange API semantics, VectorRearrange IR node generally expects shape conformance b/w vector to be permuted and index vector, since shuffle indices are held in byte array based backing storage hence compiler injects VectorLoadShuffle nodes to upcast the byte vector lanes holding indexes to match the input vector lane. Since selectFrom API already passes the indexes through vector hence we can save emitting redundant toShuffle() + toVector() operations in some cases apart from some target specific scenarios e.g. AVX2 targets [do not support direct short vector permute]instruction "VPERMW", hence we need to [massage the index vector](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L8771) to emulate desired permutation using byte permute instruction.
> 
> VectorLoadShuffle abstraction hides target specific index massaging which is why adding a target specific hook like Matcher::vector_indexes_needs_pruning compiler to selectively emit VectorLoadShuffle.

I still do not see a **definition of the semantics of RearrangeNode**:  what inputs does it accept and what does it do with them?

Can you put this explanation as comment in the code, please?

It sounds like this is what the `massaging` / `pruning` is: `emulate desired permutation using byte permute instruction.` You should find an accordingly more suiting name for the method name. Maybe it is something like `must_emulate_permutation_with...`. Or maybe it is rather a `supported` kind of question? I leave that up to you.

jatin-bhateja commented:
> > > Can you please **define** somewhere what it means to `prune indexes`? It does not help me much more than the previous "massaging indexes" you had before I asked you to change it.
> > > > Also: I'm a little worried about the semantics change of the RearrangeNode that you did with the changes in RearrangeNode::Ideal. It looks a little "hacky", especially in conjunction with the vector_indexes_needs_massaging method. Can you give a clear definition of the semantics of RearrangeNode and vector_indexes_needs_massaging, please?
> > > 
> > > 
> > > You have also not responded to this yet. It seems to me that before your proposed change, `RearrangeNode` had a clear and easy semantic, and now you somehow "hack it" to work with your `vector_indexes_needs_pruning`. Can you explain please why this makes sense and add a comment to `RearrangeNode` what its semantics is?
> > 
> > 
> > In case target does not directly support two vector selection instruction we lower the IR to its constituents, this is better than intrinsification failure as it saves costly vector boxing penalties.
> > Consider in terms of desired compiler IR and not rearrange API semantics, VectorRearrange IR node generally expects shape conformance b/w vector to be permuted and index vector, since shuffle indices are held in byte array based backing storage hence compiler injects VectorLoadShuffle nodes to upcast the byte vector lanes holding indexes to match the input vector lane. Since selectFrom API already passes the indexes through vector hence we can save emitting redundant toShuffle() + toVector() operations in some cases apart from some target specific scenarios e.g. AVX2 targets [do not support direct short vector permute]instruction "VPERMW", hence we need to [massage the index vector](https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/x86/x86.ad#L8771) to emulate desired permutation using byte permute instruction.
> > VectorLoadShuffle abstraction hides target specific index massaging which is why adding a target specific hook like Matcher::vector_indexes_needs_pruning compiler to selectively emit VectorLoadShuffle.
> 
> I still do not see a **definition of the semantics of RearrangeNode**: what inputs does it accept and what does it do with them?
> 
> Can you put this explanation as comment in the code, please?
> 
> It sounds like this is what the `massaging` / `pruning` is: `emulate desired permutation using byte permute instruction.` You should find an accordingly more suiting name for the method name. Maybe it is something like `must_emulate_permutation_with...`. Or maybe it is rather a `supported` kind of question? I leave that up to you.

Hi @eme64 , As per discussion on [PR# 20634 ](https://github.com/openjdk/jdk/pull/20634#discussion_r1759701508), we plan to suppress VectorLoadShuffle bypassing optimization for now and address this as a follow up optimization for both the flavors of selectFrom API.
I have addressed your comments. Kindly verify.

Best Regards,
Jatin

eme64 commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -2631,6 +2631,112 @@ bool LibraryCallKit::inline_vector_extract() {
+  C->set_max_vector_size(MAX2(C->max_vector_size(), (uint)(num_elem * type2aelembytes(elem_bt))));
+  return true;
+}
+
```
@jatin-bhateja You still have 3x `unbox failed v1` here. I already commented this earlier, and you resolved it and gave it a thumbs up :see_no_evil: 
Can you please fix it now?

eme64 commented on src/hotspot/share/opto/vectornode.hpp:
```diff
@@ -1612,6 +1612,19 @@ class VectorRearrangeNode : public VectorNode {
 };
 
+
+// Selects elements from second and third vector based on the indices held in
+// first vector two input vectors based on the indexes held in first vector.
+class SelectFromTwoVectorNode : public VectorNode {
+public:
+  SelectFromTwoVectorNode(Node* index, Node* src1, Node* src2, const TypeVect* vt)
+  : VectorNode(index, src1, src2, vt) {}
+
+  Node* Ideal(PhaseGVN* phase, bool can_reshape);
+  virtual int Opcode() const;
+};
```
`index` -> `indexes` because this is a vector, right? Otherwise I'll assume it is a scalar.
Can you do some pseudo-code, that says how exactly the indices are interpreted? What if they are out of bounds? Does it wrap? Or assume they are in bounds? Undefined behaviour?

eme64 replied:
For me good comments here would be immendely valuable, because it helps with other C2 optimizations.

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,70 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+
+  const TypeVect* index_vect_type = index_vec->bottom_type()->is_vect();
+  BasicType index_elem_bt = index_vect_type->element_basic_type();
+  assert(!is_floating_point_type(index_elem_bt), "");
```
Why not verify this also in the constructor of `SelectFromTwoVectorNode`? Can you maybe explicitly verify what it must be rather than **not** be?

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,70 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // are held in a byte vector which are later massaged to target specific permutation
+  // index format by subsequent VectorLoadShuffle.
+  int cast_vopc = VectorCastNode::opcode(0, index_elem_bt, true);
+  Node* index_byte_vec = phase->transform(VectorCastNode::make(cast_vopc, index_vec, T_BYTE, num_elem));
```
This cast assumes that the indices cannot have more than 8 bits. This would allow vector lengths of up to 256. This is fine for intel. But as far as I know ARM has in principle longer vectors - up to 2048 bytes. Should we maybe add some assert here to make sure we never badly truncate the index?

jatin-bhateja replied:
Shuffle overall is on our todo list, its a know limitation which we tried lifting once, yes you read it correctly, its a limitation for AARCH64 SVE once a 2048 bits vector systems are available, IIRC current max vector size on any available AARCH64 system is 256 bits, with Neoverse V2 they shrink the vector size back to 16 bytes.

eme64 replied:
Are there any asserts that would catch this?

eme64 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,70 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // Rearrange expects the indexes to lie within single vector index range [0, VLEN).
+  index_byte_vec = phase->transform(VectorNode::make(Op_AndV, index_byte_vec, bcast_lane_cnt_m1_vec, index_byte_vec->bottom_type()->is_vect()));
+
+  // Load indexes from byte vector and appropriatly massage them to target specific
+  // permutation index format.
```
I would replace `massage` -> `transform` everywhere.

eme64 reviewed (changes requested):
Looks better, I still have a few comments.

PaulSandoz commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:
```diff
@@ -2945,6 +2958,37 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
+    $abstractvectortype$ selectFrom(Vector<$Boxtype$> v1, Vector<$Boxtype$> v2);
+
+
+    /*package-private*/
```
I think we can simplify with:
```
    /*package-private*/
    @ForceInline
    final $abstractvectortype$ selectFromTemplate(Class<? extends Vector<$Boxbitstype$>> indexVecClass,
                                                  $abstractvectortype$ v1, $abstractvectortype$ v2) {
        int twoVectorLenMask = (length() << 1) - 1;
#if[FP]
        Vector<$Boxbitstype$> wrapped_indexes = this.convert(VectorOperators.{#if[intOrFloat]?F2I:D2L}, 0)
                                                   .lanewise(VectorOperators.AND, twoVectorLenMask);
        return VectorSupport.selectFromTwoVectorOp(getClass(), indexVecClass , $type$.class, $bitstype$.class,
                                                   length(), wrapped_indexes, v1, v2,
                                                   (vec1, vec2, vec3) -> selectFromTwoVectorHelper(vec1, vec2, vec3)
        );
#else[FP]
        $abstractvectortype$ wrapped_indexes = this.lanewise(VectorOperators.AND, twoVectorLenMask);
        return VectorSupport.selectFromTwoVectorOp(getClass(), indexVecClass, $type$.class, $type$.class,
                                                   length(), wrapped_indexes, v1, v2,
                                                   (vec1, vec2, vec3) -> selectFromTwoVectorHelper(vec1, vec2, vec3)
        );
#end[FP]
    }
```
(Note that's without the assert - see separate comment).

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:
```diff
@@ -2945,6 +2958,37 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
+    @ForceInline
+    final $abstractvectortype$ selectFromTemplate(Class<? extends Vector<$Boxbitstype$>> indexVecClass,
+                                                  $abstractvectortype$ v1, $abstractvectortype$ v2) {
+        int twoVectorLen = length() * 2;
```
API only accepts vector parameters and there is no means though public facing API to create a vector of NPOT sizes.

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/X-Vector.java.template:
```diff
@@ -550,6 +550,19 @@ public abstract class $abstractvectortype$ extends AbstractVector<$Boxtype$> {
+        $type$[] vecPayload3 = (($abstractvectortype$)v3).vec();
+        for (int i = 0; i < vlen; i++) {
+            int index = ((int)vecPayload1[i]);
+            res[i] = index >= vlen ? vecPayload3[index & (vlen - 1)] : vecPayload2[index];
```
Hi @PaulSandoz , we already pass wrapped indexes to this helper routine called from fallback implementation.

PaulSandoz replied:
Opps yes, the masking was throwing me off.

Can you please add a comment and/or rename the parameters e.g., so `v1` is renamed to `wrappedIndex`?
Also i would recommend not doing the masking, it is very misleading and instead do the subtraction.

eme64 reviewed (commented):
Generally, from a C2 point of view this looks good now.

sviswa7 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,69 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  // Downcast index vector to a type agnostic shuffle representation, shuffle indices
+  // are held in a byte vector which are later transformed to target specific permutation
+  // index format by subsequent VectorLoadShuffle.
+  int cast_vopc = VectorCastNode::opcode(0, index_elem_bt, true);
```
Good to use -1 when we are not sending an actual opcode:
int cast_vopc = VectorCastNode::opcode(-1, index_elem_bt, true);

sviswa7 commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -2631,6 +2632,120 @@ bool LibraryCallKit::inline_vector_extract() {
+  }
+
+  if (!arch_supports_vector(Op_SelectFromTwoVector, num_elem, elem_bt, VecMaskNotUsed)) {
+    int cast_vopc = VectorCastNode::opcode(-1, elem_bt, true);
+    if (!arch_supports_vector(Op_VectorMaskCmp, num_elem, T_BYTE, VecMaskNotUsed)            ||
+        !arch_supports_vector(Op_AndV, num_elem, T_BYTE, VecMaskNotUsed)                     ||
+        !arch_supports_vector(Op_VectorBlend, num_elem, elem_bt, VecMaskUseLoad)             ||
+        !arch_supports_vector(Op_VectorRearrange, num_elem, elem_bt, VecMaskNotUsed)         ||
+        !arch_supports_vector(cast_vopc, num_elem, T_BYTE, VecMaskNotUsed)                   ||
+        !arch_supports_vector(Op_VectorLoadShuffle, num_elem, index_elem_bt, VecMaskNotUsed) ||
+        !arch_supports_vector(Op_Replicate, num_elem, T_BYTE, VecMaskNotUsed)) {
```
We need to add VectorMaskCast here in the checks.

sviswa7 commented on src/hotspot/share/opto/vectornode.cpp:
```diff
@@ -2083,6 +2084,69 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
+  Node* lane_cnt_m1 = phase->makecon(TypeInt::make(num_elem - 1));
+  Node* bcast_lane_cnt_m1_vec = phase->transform(VectorNode::scalar2vector(lane_cnt_m1, num_elem, Type::get_const_basic_type(T_BYTE), false));
+
+  // Compute the blend mask for merging two indipendently permututed vectors
```
Typo indipendently -> independently

jatin-bhateja commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:
```diff
@@ -536,6 +536,19 @@ static ByteVector compressHelper(Vector<Byte> v, VectorMask<Byte> m) {
         return r;
     }
 
+    static ByteVector selectFromTwoVectorHelper(Vector<Byte> indexes, Vector<Byte> src1, Vector<Byte> src2) {
+        int vlen = indexes.length();
+        byte[] res = new byte[vlen];
+        byte[] vecPayload1 = ((ByteVector)indexes).vec();
+        byte[] vecPayload2 = ((ByteVector)src1).vec();
+        byte[] vecPayload3 = ((ByteVector)src2).vec();
+        for (int i = 0; i < vlen; i++) {
+            int wrapped_index = VectorIntrinsics.wrapToRange((int)vecPayload1[i], 2 * vlen);
+            res[i] = wrapped_index >= vlen ? vecPayload3[wrapped_index - vlen] : vecPayload2[wrapped_index];
+        }
+        return ((ByteVector)src1).vectorFactory(res);
+    }
+
```
> This could instead be: src1.rearrange(this.lanewise(VectorOperators.AND, 2 * VLENGTH - 1).toShuffle(), src2); Or even simplified to: src1.rearrange(this.toShuffle(), src2);

Yes, this may save additional allocation penalty of result array allocation which may slightly improve fall back performance, but logical operation cannot be directly applied over floating point vectors. so, we will need an explicit conversion to integral vector, which is why I opted for current fallback implementation which is in line with rest of the code.

sviswa7 replied:
I see the problem with float/double vectors. Let us do the rearrange form only for Integral (byte, short, int, long) vectors then. For float/double vector we could keep the code that you have currently.

jatin-bhateja replied:
You will also need additional handling for NPOT vector sizes which is handled by existing fallback implementation.

sviswa7 replied:
The intrinsic is limited to power of two. We can safely do src1.rearrange(this.lanewise(VectorOperators.AND, 2 * VLENGTH - 1).toShuffle(), src2) for integral types.

sviswa7 replied:
Agree, so we can't assume power of two in fallback.

sviswa7 commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -2631,6 +2632,179 @@ bool LibraryCallKit::inline_vector_extract() {
+  const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+
+  Node* operation = lowerSelectFromOp ?
+                      LowerSelectFromTwoVectorOperation(gvn(), opd1, opd2, opd3, vt) :
```
Thanks for bringing the lowering right here. It opens up an optimization opportunity: currently for float/double we have two casts for index (e.g. from float -> int at line 2786 and from int -> byte at line 2661 as part of LowerSelectFromTwoVectorOperation. Could this be done by one cast?

jatin-bhateja replied:
This is not sub-optimal, Float to sub-word cast is two step process where we first convert float value to integer following by integer down casting to sub-word. So resulting JIT code will still be same if we directly emit F2X or the way its handled currently. All existing targets support F2X take this route. But its good to be safe.

sviswa7 commented:
Thanks for making the changes. It looks to me that  the following checks at lines 2963-2071 in file vectorIntrinsics.cpp is now only needed when lowerSelectFromOp is false. Could you please verify and update accordingly?

 if (is_floating_point_type(elem_bt)) {
    if (!arch_supports_vector(Op_AndV, num_elem, index_elem_bt, VecMaskNotUsed)              ||
        !arch_supports_vector(cast_vopc, num_elem, index_elem_bt, VecMaskNotUsed)            ||
        !arch_supports_vector(Op_Replicate, num_elem, index_elem_bt, VecMaskNotUsed)) {
      log_if_needed("  ** index wrapping not supported: vlen=%d etype=%s" ,
                     num_elem, type2name(elem_bt));
      return false; // not supported
    }
  }

sviswa7 reviewed (approved):
Looks good to me.

sviswa7 commented:
/reviewers 2

PaulSandoz commented:
/reviewers 3

PaulSandoz commented on test/jdk/jdk/incubator/vector/templates/Unit-header.template:
```diff
@@ -397,6 +397,25 @@ relativeError));
+            for (; i < a.length; i += vector_len) {
+                for (j = 0; j < vector_len; j++) {
+                    idx = i + j;
+                    wrapped_index =(((int)order[idx]) & (2 * vector_len -1));
```
This assumes a power of two, can we change to use `Math.floorMod`?

eme64 reviewed (approved):
I gave it a quick scan, and I have no further comments. LGTM.

PaulSandoz commented:
> > I gave it a quick scan, and I have no further comments. LGTM.
> 
> Thank you, i will kick off an internal test.

Tier 1 to 3 tests pass.

jatin-bhateja commented:
/integrate

jatin-bhateja commented:
Thanks @PaulSandoz , @sviswa7 and @eme64 for review suggestions.

sviswa7 reviewed (approved):
Changes look good to me.

jatin-bhateja commented:
/integrate

liach commented:
This patch failed on the lastest master. Another reason OpenJDK guide asks to merge master despite all these commit churns...
